src/cudadecoder/batched-threaded-nnet3-cuda-online-pipeline.cc

```diff
@@ -43,9 +43,10 @@ BatchedThreadedNnet3CudaOnlinePipeline::~BatchedThreadedNnet3CudaOnlinePipeline(
 ) {
   // The destructor races with callback completion. Even if all callbacks have
   // finished, the counter may (non-deterministically) lag behind by a few ms.
+  // Deleting the object when all callbacks had been called is UB: the variable
+  // n_lattice_callbacks_not_done_ is accessed after a callback has returned.
   WaitForLatticeCallbacks();
   KALDI_ASSERT(n_lattice_callbacks_not_done_ == 0);
-  KALDI_ASSERT(!cuda_fst_ || cuda_fst_.use_count() == 1);
```

Comment:
In the latest version (#4548), I changed it to unique_ptr, so the count is no more. It was only passed to the decoder, exactly once, anyway, and as a raw pointer.

Reply:
Let me pre-squash separate commits on this branch and push it clean, so that line like this aren't confusing. I added and removed the assert, because in the other PR #4548 I replaced the shared_ptr with a unique_ptr.